### PR TITLE
Refactor of `model_function_signature()` to address complexity limits.

### PR DESCRIPTION
**Pull request by peterschutt**
- Should be no logical changes.
- Breaks out detailed logic into methods.
- High-level business logic in `.model()` method.
- Added a type to hold interesting properties of the parameters as we iterate over them.
- Docstrings everywhere.

### CODE REVIEW

File: starlite/signature.py [hunk 1]
```diff
@@ -60,63 +76,247 @@ def parse_values_from_connection_kwargs(
...
+    @property
+    def default_defined(self) -> bool:
+        """
+        `True` if `self.default` is not one of the undefined sentinel types.
+
+        Returns
+        -------
+        bool
+        """
+        return self.default not in {Signature.empty, Undefined}
```
Comment (Goldziher) [inline starlite/signature.py:99]: Predeclare the set outside the class, it will make things faster.

File: starlite/signature.py [hunk 2]
```diff
@@ -60,63 +76,247 @@ def parse_values_from_connection_kwargs(
...
-            if default not in [signature.empty, Undefined]:
-                field_definitions[kwarg] = (type_annotation, default)
-                defaults[kwarg] = default
-            elif not type_optional:
-                field_definitions[kwarg] = (type_annotation, ...)
-            else:
-                field_definitions[kwarg] = (type_annotation, None)
-        model: Type[SignatureModel] = create_model(
-            fn_name + "_signature_model", __base__=SignatureModel, **field_definitions
+    name: str
```
Comment (Goldziher) [inline starlite/signature.py:85]: Please add a `__slots__` dunder

File: starlite/signature.py [hunk 3]
```diff
@@ -60,63 +76,247 @@ def parse_values_from_connection_kwargs(
...
+    field_definitions : dict[str, Tuple[Any, Any]
+        Maps parameter name to the `(<type>, <default>)` tuple passed to `pydantic.create_model()`.
+    defaults : dict[str, Any]
+        Maps parameter name to default value, if one defined.
+    dependency_name_set : set[str]
+        The names of all known dependency parameters.
+    """
+
+    # names of fn params not included in signature model.
+    SKIP_NAMES = {"self", "cls"}
```
Comment (Goldziher) [inline starlite/signature.py:159]: Also here add a __slots__ dunder please

File: starlite/signature.py [hunk 4]
```diff
@@ -60,63 +76,247 @@ def parse_values_from_connection_kwargs(
...
+        signature that should be included in the `SignatureModel` type.
+
+        Returns
+        -------
+        Generator[SignatureParameter, None, None]
+        """
+        for name, parameter in self.signature.parameters.items():
+            if name in self.SKIP_NAMES:
+                continue
+            yield SignatureParameter.new(self.fn_name, name, parameter)
```
Comment (Goldziher) [inline starlite/signature.py:284]: Why use new rather than a standard constructor?

**Review by Goldziher — APPROVED** [review]

File: starlite/signature.py (hunk 4 above)
Comment (peterschutt) [reply starlite/signature.py:284]: If I think of a good reason I'll let you know

**Pull request merged**